# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [27]:
import pandas as pd
import numpy as np

## Input Dataset

In [28]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [29]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [30]:
df.head()

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0          1    15634602    France  Female   42       2       0.00   
1          2    15647311     Spain  Female   41       1   83807.86   
2          3    15619304    France  Female   42       8  159660.80   
3          4    15701354    France  Female   39       1       0.00   
4          5    15737888     Spain  Female   43       2  125510.82   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0

In [31]:
df.shape

(10000, 12)

In [32]:
df.nunique()

RowNumber          10000
CustomerId         10000
Geography              3
Gender                 2
Age                   70
Tenure                11
Balance             6382
NumOfProducts          4
HasCrCard              2
IsActiveMember         2
EstimatedSalary     9999
Exited                 2
dtype: int64

In [33]:
df['Gender'].unique()

array(['Female', 'Male'], dtype=object)

In [34]:
df['Geography'].unique()

array(['France', 'Spain', 'Germany'], dtype=object)

In [35]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

# Hilangkan kolom yang tidak relevan
kolom_yang_penting = ['Geography', 'Gender', 'Age', 'Tenure',
                     'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
                     'EstimatedSalary', 'Exited']

df = df[kolom_yang_penting].copy()

In [36]:
df.columns

Index(['Geography', 'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts',
       'HasCrCard', 'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')

## Preprocessing

In [37]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df, columns=['Geography', 'Gender'])

In [38]:
df.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Exited  Geography_France  Geography_Germany  \
0        101348.88       1                 1                  0   
1        112542.58       0                 0                  0   
2        113931.57       1                 1                  0   
3         93826.63       0                 1                  0   
4         79084.10       0                 0                  0   

   Geography_Spain  Gender_Female  Gender_Male  
0                0              1            0  
1                1              1            0  
2                0              1            0  
3                0              1            0  
4                1              1            0

In [39]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
# Pisahkan fitur (X) dan target (Y)
X = df.drop('Exited', axis=1)  # Mengambil semua kolom kecuali 'Exited'
y = df['Exited']  # Mengambil kolom 'Exited' sebagai target (Y)

In [40]:
X.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Geography_France  Geography_Germany  Geography_Spain  \
0        101348.88                 1                  0                0   
1        112542.58                 0                  0                1   
2        113931.57                 1                  0                0   
3         93826.63                 1                  0                0   
4         79084.10                 0                  0                1   

   Gender_Female  Gender_Male  
0              1            0  
1              1            0  
2              1            0  
3              1            0  
4              1            0

In [41]:
y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [42]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)


In [43]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()

In [44]:
X_transform.head()

Age  Tenure   Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0  0.324324     0.2  0.000000       0.000000        1.0             1.0   
1  0.310811     0.1  0.334031       0.000000        0.0             1.0   
2  0.324324     0.8  0.636357       0.666667        1.0             0.0   
3  0.283784     0.1  0.000000       0.333333        0.0             0.0   
4  0.337838     0.2  0.500246       0.000000        1.0             1.0   

   EstimatedSalary  Geography_France  Geography_Germany  Geography_Spain  \
0         0.506735               1.0                0.0              0.0   
1         0.562709               0.0                0.0              1.0   
2         0.569654               1.0                0.0              0.0   
3         0.469120               1.0                0.0              0.0   
4         0.395400               0.0                0.0              1.0   

   Gender_Female  Gender_Male  
0            1.0          0.0  
1            1.0          0.0  
2            1.0          0.0  
3            1.0          0.0  
4            1.0          0.0

## Train-Test Split

In [45]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

In [46]:
X_train.shape, y_train.shape

((7500, 12), (7500,))

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

* Regresi logistik adalah teknik analisis data yang menggunakan matematika untuk menemukan hubungan antara dua faktor data. Kemudian menggunakan hubungan ini untuk memprediksi nilai dari salah satu faktor tersebut berdasarkan faktor yang lain. Prediksi biasanya memiliki jumlah hasil yang terbatas, seperti ya atau tidak.

In [47]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}

#START CODE
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV

model1 = LogisticRegression()

params = {
    'tol': [0.1, 0.01, 0.001],
    'C': [0.5, 1.0, 1.5, 2.0]
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
grid = GridSearchCV(
    estimator=model1,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,
    cv=10
)

grid.fit(X_train,y_train)

best_params = grid.best_params_
print("Parameter terbaik:", best_params)

Parameter terbaik: {'C': 0.5, 'tol': 0.1}


In [48]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.97      0.89      1983
           1       0.66      0.20      0.31       517

    accuracy                           0.81      2500
   macro avg       0.74      0.59      0.60      2500
weighted avg       0.79      0.81      0.77      2500


[[1929   54]
 [ 412  105]]

0.8136


## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

* Metode XGBoost merupakan pengembangan dari gradient boosting yang diusulkan oleh Dr. Tianqi Chen dari University of Washington pada tahun 2014 [18]. Gradient boosting merupakan algoritma yang dapat menemukan solusi yang optimal untuk berbagai masalah khususnya pada regresi, klasifikasi dan ranking.

In [23]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}

#START CODE
import xgboost as xgb
from sklearn.model_selection import GridSearchCV

model2 = xgb.XGBClassifier()

params = {
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 4, 5],
    'n_estimators': [100, 200, 300],
    'subsample': [0.8, 0.9, 1.0],
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
grid = GridSearchCV(
    estimator=model2,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,
    cv=10
)

grid.fit(X_train,y_train)

best_params = grid.best_params_
print("Parameter terbaik:", best_params)


Parameter terbaik: {'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 200, 'subsample': 0.8}


In [24]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.96      0.92      1983
           1       0.77      0.50      0.61       517

    accuracy                           0.87      2500
   macro avg       0.83      0.73      0.77      2500
weighted avg       0.86      0.87      0.86      2500


[[1907   76]
 [ 256  261]]

0.8672


## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

* Support Vector Machine (SVM) merupakan sebuah algoritma klasifikasi untuk data linear dan non-linear. SVM menggunakan mapping non-linear untuk mentransformasikan training data awal ke dimensi yang lebih tinggi.

In [25]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}

#START CODE
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

model3 = SVC()

params = {
    'tol': [0.1, 0.01, 0.001],
    'C': [0.5, 1.0, 1.5, 2.0]
}
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
grid = GridSearchCV(
    estimator=model3,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,
    cv=10
)

grid.fit(X_train,y_train)

best_params = grid.best_params_
print("Parameter terbaik:", best_params)


Parameter terbaik: {'C': 2.0, 'tol': 0.01}


In [26]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.84      0.99      0.91      1983
           1       0.86      0.29      0.43       517

    accuracy                           0.84      2500
   macro avg       0.85      0.64      0.67      2500
weighted avg       0.85      0.84      0.81      2500


[[1958   25]
 [ 367  150]]

0.8432


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

* Dengan mengacu pada nilai akurasi yang tercapai, dapat disimpulkan bahwa model XGBoost dengan hyperparameter grid (0.8776) merupakan yang terbaik di antara tiga model yang telah digunakan. Penilaian ini didasarkan pada fakta bahwa model XGBoost berhasil mencapai tingkat akurasi yang lebih tinggi dibandingkan dengan model Logistik Regresi (0.8136) dan SVM (0.8632) dengan berbagai konfigurasi parameter yang telah diuji. Alasannya adalah algoritma ensemble yang sangat kuat dan sering kali mengungguli model-model lain dalam berbagai tugas Machine Learning, terutama ketika telah mengalami penyetelan hyperparameter yang cermat. Tingkat akurasi yang lebih tinggi ini menggambarkan kemampuan model XGBoost yang lebih baik dalam melakukan prediksi terhadap data yang digunakan.